lib/ansible/module_utils/network/nxos/utils/telemetry/telemetry.py

```diff
@@ -0,0 +1,186 @@
+            mp[key] = module_params[key]
+
+    if tms_config == 'TMS_DESTGROUP':
+        mp['destination_groups'] = []
+        for destgrp in module_params['destination_groups']:
+            if destgrp['id'] == resource_key:
+                mp['destination_groups'].append(destgrp)
+
+    if tms_config == 'TMS_SENSORGROUP':
+        mp['sensor_groups'] = []
+        for sensor in module_params['sensor_groups']:
+            if sensor['id'] == resource_key:
+                mp['sensor_groups'].append(sensor)
+
+    if tms_config == 'TMS_SUBSCRIPTION':
+        mp['subscriptions'] = []
+        for sensor in module_params['subscriptions']:
+            if sensor['id'] == resource_key:
+                mp['subscriptions'].append(sensor)
+
+    return mp
+
+
+def valiate_input(playvals, type, module):
+    """
+    Helper method to validate playbook values for destination groups
+    """
+    if type == 'destination_groups':
+        if not playvals.get('id'):
+            msg = "Invalid playbook value: {0}.".format(playvals)
+            msg = msg + " Parameter <id> under <destination_groups> is required"
+            module.fail_json(msg=msg)
+        if playvals.get('destination') and not isinstance(playvals['destination'], dict):
+            msg = "Invalid playbook value: {0}.".format(playvals)
+            msg = msg + " Parameter <destination> under <destination_groups> must be a dict"
+            module.fail_json(msg=msg)
+        if not playvals.get('destination') and len(playvals) > 1:
+            msg = "Invalid playbook value: {0}.".format(playvals)
+            msg = msg + " Playbook entry contains unrecongnized parameters."
+            msg = msg + " Make sure <destination> keys under <destination_groups> are specified as follows:"
+            msg = msg + " destination: {ip: <ip>, port: <port>, protocol: <prot>, encoding: <enc>}}"
+            module.fail_json(msg=msg)
+
+    if type == 'sensor_groups':
+        if not playvals.get('id'):
+            msg = "Invalid playbook value: {0}.".format(playvals)
+            msg = msg + " Parameter <id> under <sensor_groups> is required"
+            module.fail_json(msg=msg)
+        if playvals.get('path') and 'name' not in playvals['path'].keys():
+            msg = "Invalid playbook value: {0}.".format(playvals)
+            msg = msg + " Parameter <path> under <sensor_groups> requires <name> key"
+            module.fail_json(msg=msg)
+
+
+def get_instance_data(key, cr_key, cr, existing_key):
+    """
+    Helper method to get instance data used to populate list structure in config
+    fact dictionary
+    """
+    data = {}
+    if existing_key is None:
+        instance = None
+    else:
+        instance = cr._ref[cr_key]['existing'][existing_key]
+
+    if key == 'destination_groups':
+        m = re.search(r"destination-group (\d+)", cr._ref['_resource_key'])
+        instance_key = m.group(1)
+        data = {'id': instance_key, cr_key: instance}
+
```

Comment:
Part 1. Consider... this code looks identical except for the regex patterns so you could just remove the `if` blocks and do this to handle all cases:

```
    patterns = {
        'destination_groups': r"destination-group (\d+)",
        ...
    }
    m = re.search(patterns[key], cr._ref['_resource_key'])
    instance_key = m.group(1)
    data = {'id': instance_key, cr_key: instance}
```

Part 2. You're not checking if the match `m` was populated so `m.group(1) would fail for a regex non-match. Are you guaranteed to always have a match for these entries?

Reply:
👍 on Part 1.  For Part 2... yes, we are guaranteed to always have a match.

Reply:
👍